library/core/src/fmt/num.rs

```diff
@@ -459,5 +498,176 @@ mod imp {
+        }
+
+        // `rem` < 1e4 < (1 << 16)
+        let mut rem = rem as u16;
+        if rem >= 100 {
+            let d1 = (rem % 100) << 1;
+            rem /= 100;
+            *curr -= 2;
+            ptr::copy_nonoverlapping(lut_ptr.offset(d1 as isize), buf_ptr.offset(*curr), 2);
+        }
+
+        // decode last 1 or 2 chars
+        if rem < 10 {
+            *curr -= 1;
+            *buf_ptr.offset(*curr) = (rem as u8) + b'0';
+        } else {
+            let d1 = rem << 1;
+            *curr -= 2;
+            ptr::copy_nonoverlapping(lut_ptr.offset(d1 as isize), buf_ptr.offset(*curr), 2);
+        }
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl fmt::Display for u128 {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt_u128_2(*self, true, f)
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl fmt::Display for i128 {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let is_nonnegative = *self >= 0;
+        let n = if is_nonnegative {
+            self.to_u128()
+        } else {
+            // convert the negative num to positive by summing 1 to it's 2 complement
+            (!self.to_u128()).wrapping_add(1)
+        };
+        fmt_u128_2(n, is_nonnegative, f)
+    }
+}
+
+#[allow(dead_code)]
+fn fmt_u128_2(n: u128, is_nonnegative: bool, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    // 2^128 is about 3*10^38, so 39 gives an extra byte of space
+    let mut buf = [MaybeUninit::<u8>::uninit(); 39];
+    let mut curr = buf.len() as isize;
+    let buf_ptr = MaybeUninit::first_ptr_mut(&mut buf);
+
+    // SAFETY: Since `d1` and `d2` are always less than or equal to `198`, we
+    // can copy from `lut_ptr[d1..d1 + 1]` and `lut_ptr[d2..d2 + 1]`. To show
+    // that it's OK to copy into `buf_ptr`, notice that at the beginning
+    // `curr == buf.len() == 39 > log(n)` since `n < 2^128 < 10^39`, and at
+    // each step this is kept the same as `n` is divided. Since `n` is always
+    // non-negative, this means that `curr > 0` so `buf_ptr[curr..curr + 1]`
+    // is safe to access.
+    unsafe {
+        let (n, rem) = udiv_1e9(n);
+        fmt_u64_2(rem, &mut buf, &mut curr);
+
+        if n != 0 {
+            let (n, rem) = udiv_1e9(n);
+            fmt_u64_2(rem, &mut buf, &mut curr);
+            // Should this following branch be annotated with unlikely?
+            if n != 0 {
+                curr -= 1;
+                *buf_ptr.offset(curr) = (n as u8) + b'0';
+            }
+        }
+    }
+
+    // SAFETY: `curr` > 0 (since we made `buf` large enough), and all the chars are valid
+    // UTF-8 since `DEC_DIGITS_LUT` is
+    let buf_slice = unsafe {
+        str::from_utf8_unchecked(slice::from_raw_parts(
+            buf_ptr.offset(curr),
+            buf.len() - curr as usize,
+        ))
+    };
+    f.pad_integral(is_nonnegative, "", buf_slice)
+}
+
+fn udiv_1e9(n: u128) -> (u128, u64) {
+    const DIV: u64 = 1e19 as u64;
```

Comment:
You can remove this hard "as" cast from your code like this (needs const_int_pow feature):

    const DIV: u64 = 10_u64.pow(19);

Reply:
Do you mean this specific one or all the `as`casts? In the other cases, I inlined the others, and while technically it should be const-propagated, the cast to me should always occur at compile time, more so than doing `.pow(..)`.

Reply:
I meant this specific cast.

(I'd like to see Rustc devs and stdlib devs start getting into the habit of minimizing the number of has "as" casts in the rustc codebase, because such casts are sharp blades that sometimes cut you.)

Regarding the performance, using a const like I've suggested is zero-cost at run-time (I think).

Reply:
In what case can these casts cause unexpected behaviour? In this case the constant is definitely in the u64 range.

If it's defined as `const VAR: u64 = { expr }`, then there is definitely no runtime cost. Inlining it elsewhere might incur some runtime cost.

Reply:
>In what case can these casts cause unexpected behaviour? In this case the constant is definitely in the u64 range.<

(That specific "as" doesn't cause unexpected behaviour. But being willing to use "as" everywhere is like being against the presence of unsafe{} statement in Rust language on the base of the code inside a specific usage of unsafe{} in a program being evidently safe. Even if a specific usage of "as" is safe, it's a sharp tool, and it may lead to problems in other cases, so better to minimize its usage in a codebase. I can call this language blindness: what a language puts in front is visible and taken care of, what a language doesn't care about the programmer too doesn't care regardless the troubles it could cause. But eventually the unsafety of using "as" will come out in Rust culture. The recently introduced Rustc suggestions to use try_into are a step forward).